rfcs/002-runtime-module-management/rfc.md

```diff
@@ -0,0 +1,654 @@
+- **loaded**: an boolean value indicating whether this module is loaded
+- **error**: an Error instance if there is an error during the initialization of this module or `null` if initialization succeed
+
+The definition of modules:
+
+```ts
+interface Module {
+  id: string;
+  exports: any;
+  meta: Record<string, any>;
+  loaded: boolean;
+  error?: Error;
+}
+```
+
+#### Module initialization function
+
+A module will be transformed to a **sync** or an **async** module initialization function, its body is the module content, when a module is `import`ed/`require`d, the module initialization function will be called.
+
+Module initialization functions take three arguments:
+
+- **`module`**: the module instance
+- **`exports`**: the exports object of the module
+- **`__farm_require__`**: the require function, used to require other modules
+
+Module initialization function has no return value, after the function is executed, the `exports` object should contain all exported bindings of the module, or an error will be thrown if something go wrong during initialization.
+
+The definition of module initialization functions is:
+
+```ts
+type ModuleInitialization = (module: Module, exports: Record<string, any>, __farm_require__: (request: string) => Module);
+```
+
+Below is an example:
+
+```js
+// foo.js
+export const foo = 1;
+
+// transformed runtime module
+const modules = {
+  'foo.js': /* runtime module starts here */function(module, exports, __farm_require__) {
+    exports.foo = 1;
+  } /* runtime module ends here */
+}
+```
+
+#### Module system architecture
+
+Module system consists of three parts:
+
+- **module initialization functions map**: an object containing all initialization functions of modules
+- **cached modules map**: an object containing all cached modules
+- **loaded resources map**: an object containing all resources' id which have been loaded (only for legacy browsers)
+
+```ts
+// pseudo code
+class ModuleSystem {
+  readonly moduleMaps: Record<string, ModuleInitialization>;
+  readonly cacheMaps: Record<string, Module>;
+  readonly loadedResourceMap: Record<string, boolean>;
+
+  require(id: string): Module {}
+  register(id: string, moduleInitialization: ModuleInitialization): void {}
+  update(id: string, moduleInitialization: ModuleInitialization): void {}
+  delete(id: string): boolean {}
+  deleteCache(id: string): boolean {}
+}
+```
+
+#### Module system API
+
+Here is the APIs exposed by module system:
+
+```ts
+interface ModuleSystem {
+  require(id: string): Module;
+  register(id: string, moduleInitialization: ModuleInitialization): void;
+  update(id: string, moduleInitialization: ModuleInitialization): void;
+  delete(id: string): boolean;
+  deleteCache(id: string): boolean;
+}
+```
+
+#### Module system API alias
+
+For a smaller production size, every API has an short-name alias:
+
+- `require` → `r`
+- `register` → `g`
+- `update` → `u`
+- `delete` → `d`
+- `deleteCache` → `c`
+
+#### Module states
```

Comment:
This state will not be saved, it's used to describe the state change for a module.

If a module is registered in module initialization map, it correspond to the state `registered`, if it is initialized, it will turn to either `initialized` or `failed`, the `.loaded` and `.error` property of module indicate the state.